### ISSUE TITLE
[v2 FEATURE]: Report test set performance at the end of `hpopt` and `train` 

### ISSUE THREAD

**Issue by spyda90** [chemprop/chemprop#1260]
Hi,

Sorry for pushin multiple request and questions, but I am strongly convinced that while testing multiple hpopt/training settings, generating test_score file is crucial (as it ws in the v1). In v2 test predictions are written, that can be good but not sufficient in my opinion.

I don't know if I missed some options that allow to do that. I fthat is the case,please notify it to me.
However, it would be great to have a similar feature in v2. This would allow to compare the performance within the test set for different settings.

Thank you in advance

**Comment by JacksonBurns** [comment]
Chemprop v1 actually had some unintended behavior where hyperparameters would be set based on the performance on the test set (in code [here](https://github.com/chemprop/chemprop-v1-old-branches/blob/97c03cb658be4081bbc4e8655e6670eff80d26a4/chemprop/train/run_training.py#L411)). Chemprop v2 fixes this by partitioning out a test set, but then [not using it](https://github.com/chemprop/chemprop/blob/main/chemprop/cli/hpopt.py#L540) in favor of the val dataset (as it should).

We could write the performance on the witheld test set for the final optimized model, if that would be helpful? What are you trying to do with the test performance?

**Comment by JacksonBurns** [comment]
Internally chemprop uses the performance on the validation set to determine the best configuration, and then these scores for all the different configurations are written to a CSV file:
https://github.com/chemprop/chemprop/blob/c4602da66bf9dc766ea48af126fc2600b445e923/chemprop/cli/hpopt.py#L613

**Comment by JacksonBurns** [comment]
Yeah, we can add that. I'll open a PR

**Comment by JacksonBurns** [comment]
@spyda90 so you want to see the test performance during training? This is already printed at the end of training. Is it that you want to see the individual predictions?

**Comment by spyda90** [comment]
Hi Jackson,

Currently as training output files in V2 I can see (in any replicate or ensemble model) a` test_predictions.csv` that contains individual predictions of the test set.
However, we'd like that, at the end of the training process, a second file would be generated, containing the performance on the test set (`test_scores.csv`, as it was in the v1).

It would be even better if:
- in case of ensemble only one file is generated (as mean of the metrics) (V1 file attached)
- in case of replicates, a summary file of performance is generated (including mean and standard deviation as well as single replicates metrics, as it was in case of cross-validation for the v1) (V1 fil attached, CV)

[test_scores.csv](https://github.com/user-attachments/files/20981933/test_scores.csv)
[test_scores_CV.csv](https://github.com/user-attachments/files/20981934/test_scores_CV.csv)
